Type completeness in extend (#6395)
carbon-language/carbon-lang: proposals/p6395.md

<!-- toc -->

## Table of contents

-   [Abstract](#abstract)
-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Details](#details)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)

<!-- tocstop -->

## Abstract

Require the facet type in `extend require impls` and `extend impl as`
declarations to be complete, since `extend` changes the scopes which are looked
in for name lookup.

## Problem

Proposal
[#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md)
laid out rules for when a facet type needs to be identified or complete. When
`require impls Z` is written, then `Z` must be identified. However it does not
specify any rule for `extend require impls Z`. or `extend impl as Z`.

If an interface `Y` uses `extend` for another interface `Z`, then a facet type
for `Y` will also allow name lookup into `Z`. Similarly if a class `Y` uses
`extend` for an interface `Z`, then an object of type `Y` will also allow name
lookup into `Z`. In both cases, when the definition of `Y` is complete, we
should also require `Z` to be complete, as its name scope becomes conceptually a

> part of `Y`.

Reviewer: For completeness, consider a interface that extends itself:
```
interface I(T:! type) {
  extend require impls I(());
}
```
`I` is not complete at the point of the `extend require` declaration, but name lookup into `I` should be possible by the end of its definition. Of course, all the names are going to conflict (unless `T` is `()`), so this may not be a case worth worrying too much about.

Author: Yeah I think this is okay to exclude by requiring `I` to be complete. Explicitly say so at the end of details now.

    Consistent class and interface syntax

## Proposal

Require that `extend require impls Z` requires `Z` to be complete where the
declaration is written. Similarly, require that `extend impl as Z` requires `Z`
to be complete where the declaration is written. In both cases, this ensures
that `Z` is complete by the time the definition of the enclosing class,
interface, or named constraint is complete. And it ensures that we can perform
name lookup into `Z` from within the enclosing scope of the `extend` declaration
thereafter.

We call an `extend require` declaration _extend-reachable_ from a facet type if
it's enclosing interface or named constraint is part of that facet type, or if
its enclosing interface or named constraint is named by another
_extend-reachable_ declaration. That is, we can find a path from the facet type
to the `extend require` declaration through other `extend require` declarations.

We say that a facet type is complete if:

-   All its _extend-reachable_ `extend require` declarations are complete. This
    follows from the enclosing interface or named constraint being complete.
-   We are able to form a specific for each interface and named constraint

>     extended by the facet type.

Reviewer: What would prevent forming a specific in this situation? Can you give an example? Or is that about the monomorphization errors you mention in the details section?

Author: It is the monomorphization errors, yes. Otherwise, checked generics protect against this. But we don't have a facet capability to guard everything.

I added this to the line.

-   We are able to form a specific for each _extend-reachable_ `extend require`
    declaration in the context of the specific facet type being completed.

## Details

To `extend` an entity `Y` with another `Z` means that name lookups into `Y` will
also look into `Z`. At the end of the definition of `Y`, we expect to be able to
perform name lookup into `Y`, as it is complete. To maintain this property
through `extend`, we require that `extend require` declarations only name other
interfaces or named constraints that are themselves complete.

This functions recursively: Since each interface or named constraint is
complete, we already know the `extend require` declarations inside are complete.
However it is possible to uncover monomorphization errors once a specific is
applied to the facet type in an `extend require` or `extend impls` declaration.
For example:

```carbon
interface Z(T:! type) {}

class C(N! i32) {
  extend impls Z(array(i32, N)) {}
}
```

> Here the facet yype `Z(array(i32, N))` is complete at the point where it is

Reviewer 2: ```suggestion
Here the facet type `Z(array(i32, N))` is complete at the point where it is
```

Author: Done

# Type completeness in extend

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

[Pull request](https://github.com/carbon-language/carbon-lang/pull/6395)

<!-- toc -->

## Table of contents

-   [Abstract](#abstract)
-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Details](#details)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)

<!-- tocstop -->

## Abstract

Require the facet type in `extend require impls` and `extend impl as`
declarations to be complete, since `extend` changes the scopes which are looked
in for name lookup.

## Problem

Proposal
[#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md)
laid out rules for when a facet type needs to be identified or complete. When
`require impls Z` is written, then `Z` must be identified. However it does not

> specify any rule for `extend require impls Z`. or `extend impl as Z`.

Reviewer 2: ```suggestion
specify any rule for `extend require impls Z` or `extend impl as Z`.
```

Author: Done

If an interface `Y` uses `extend` for another interface `Z`, then a facet type
for `Y` will also allow name lookup into `Z`. Similarly if a class `Y` uses
`extend` for an interface `Z`, then an object of type `Y` will also allow name
lookup into `Z`. In both cases, when the definition of `Y` is complete, we
should also require `Z` to be complete, as its name scope becomes conceptually a
part of `Y`.

Additionally, after an `extend ... Z` declaration inside the definition of `Y`,
the names of `Z` should be immediately available inside the `Y` definition,
including inside the nested definition of `extend impl as Z { ... }`.

## Background

-   Proposal
    [#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md):
    Forward impl declaration of an incomplete interface
-   Proposal
    [#2760](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2760.md):
    Consistent class and interface syntax

> ## Proposal

Reviewer 2: Do we already have a rule for `extend base` and `extend adapt` declarations? If not, would it be worth adding them here too, so that we have complete coverage of all `extend` declarations?

to be complete where the declaration is written. In both cases, this ensures
that `Z` is complete by the time the definition of the enclosing class,
interface, or named constraint is complete. And it ensures that we can perform
name lookup into `Z` from within the enclosing scope of the `extend` declaration
thereafter.

We call a `extend require` declaration _extend-reachable_ from a facet type if
it's enclosing interface or named constraint is part of that facet type, or if
its enclosing interface or named constraint is named by another
_extend-reachable_ declaration. That is, we can find a path from the facet type
to the `extend require` declaration through other `extend require` declarations.

We say that a facet type is complete if:

-   All its _extend-reachable_ `extend require` declarations are complete. This
    follows from the enclosing interface or named constraint being complete.
-   We are able to form a specific for each interface and named constraint
    extended by the facet type.
-   We are able to form a specific for each _extend-reachable_ `extend require`
    declaration in the context of the specific facet type being completed,

>     without monomorphization errors.

Reviewer 2: It'd be nice if we could say something broader here, and while I don't think we can avoid facet types being a special case, perhaps we can make them reuse the broader notion. What I'm thinking is something along the lines of:

"""
An `extend` declaration

```
extend base: X;
extend adapt X;
extend impl as X;
extend require impls X;
```

declares that the enclosing scope _extends_ the scope associated with `X`. The extend declaration requires `X` to be complete, or, if `X` is a generic parameter, requires the type of `X` to be complete.

The scope of a facet type formed by a `where` declaration extends the scope of its first operand. The scope of a facet type formed by an `&` operator extends the scopes of both of its operands. In either case, the facet type is complete if every scope it extends is complete. The facet type `type` is associated with an empty scope and is complete.
"""

Author: Sounds good, I could not find anything else written down for other extend decls. I took a lot of this text verbatim (but I wanted to call it "extended by" rather than "extends", as that direction felt better?

## Details

To `extend` an entity `Y` with another `Z` means that name lookups into `Y` will

> also look into `Z`. At the end of the definition of `Y`, we expect to be able to

Reviewer 2: I would expect this to apply immediately, not at the end of the definition of `Y`. For example:
```carbon
base class Z {
   class X {}
}
class Y {
  extend base: Z;
  // Finds `X` in `Z`.
  fn F() -> X;
}
```

Author: Yeah, me too. In fact this is meant to come with tests showing that expectation (that got lost in that deleted file mistake). I thought I had added that as an additional step later in the proposal but I don't see it, and there's no need to hedge up here, so yeah let me fix that.

for other `extend` declarations, including `extend impl as X`, `extend base: X`,
and `extend adapt X`.

An `extend` declaration always names one or more target entities which will be
included in the search for names when looking in the enclosing scope of the
`extend` declaration. In order to do name lookup into an entity, it must be
complete to avoid poisoning names in the entity.

## Background

-   Proposal
    [#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md):
    Forward `impl` declaration of an incomplete interface
-   Proposal
    [#2760](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2760.md):
    Consistent `class` and `interface` syntax
-   Proposal
    [#0777](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0777.md):
    Inheritance

## Proposal

An `extend` declaration declares that the enclosing scope is extended by the
scope associated with any target scope named in the declaration. The `extend`
declaration requires that the target scopes named in an `extend` declaration are
complete, or if the target is a generic parameter, requires the type of the

> parameter to be complete.

Reviewer 2: You're using "is extended by" in the opposite sense from the way around that I would expect and that the syntax suggests. If `B` contains `extend ... A`, then I would say that "B extends A", and I think that makes sense as the scope of B is larger than the scope of A (B's scope is an extension of A's scope). But here you're instead saying that if `B` contains `extend A` then "B is extended by A", which presumably means that we could equally say "A extends B'. To me that seems backwards.

Author: I guess it's just a matter of perspective. I see `B` contains `extend...A` as meaning that the scope of `B` is extended (via `AddExtendedScope`) from its original scope to a larger one that includes `A`. But giving `A` any agency here to say it "extends" anything wouldn't make sense in that formulation, since `B` is the one in control and acting, through its `extend` decl.

Anyhow I am fine with swapping things given this didn't land with you.

| ---------------------------- | ------------------------------------- |
| `I`                          | Requires `I` is complete.             |
| `I & J`                      | Requires `I` and `J` are complete.    |
| `type where U impls J`       | Requires `type` is complete.          |
| `I & (type where U impls J)` | Requires `I` and `type` are complete. |

## Details

To `extend` an entity `Y` with another `Z` means that name lookups into `Y` will
also look into `Z`. Immediately after the `extend` operation, members of `Z`
should also be found when doing name lookup into `Y`, both from outside and from
inside the definition of `Y`. In order to be able to perform lookups into `Z`,
we require that `extend` operations only target scopes that are complete.

This requirement functions recursively: Given an interface `A` that is extended
by another interface `B`. By naming `A` in an extend declaration, we require `A`
is complete. This provides that its entire definition is known, and thus its
`extend` relationship to `B`. The `extend` relationship there also provides that
`B` is complete.

If the target scope of an `extend` declaration is a generic parameter, its type

> must be complete where the `extend` declaration is written. As the generic

Reviewer 2: Might be worth saying here that the reason for this is that name lookups into the extended scope will look into the type of the generic parameter. Eg:
```
interface I {
  fn F();
}
class C(T:! I) {
  extend base: T;
  // `F` names `T.F` here, found in `I`.
  fn G() { F(); }
}
```

Author: Done

## Proposal

An `extend` declaration declares that the enclosing scope extends the scope

> associated with any target scope named in the declaration. That is to say name

Reviewer 2: Maybe:
```suggestion
An `extend` declaration declares that the enclosing scope extends the scope
associated with the target entity named in the declaration. That is to say name
```
"scopes associated with a scope" sounds a bit strange to me. Also, for example `extend impl as I(X)` names both `I(X)` and `X`, and "any target scope named in the declaration" might suggest we extend both!

Author: Right, the issue with "the entity" to me is that facet types aren't entities but they can name more than one entity. So maybe "any target entity".

lookups into the enclosing scope will also look into the scopes which are

> extended by the `extend` declaration. The `extend` declaration requires that the

Reviewer 2: ```suggestion
lookups into the enclosing scope will also look into the scopes which are
nominated by the `extend` declaration. The `extend` declaration requires that the
```
(With the terminology switch, the scope being extended is the enclosing scope of the extend declaration.)

inside the definition of `Y`. In order to be able to perform lookups into `Z`,
we require that `extend` operations only target scopes that are complete.

This requirement functions recursively. Given an interface `A` that extends

> another interface `B`: By naming `A` in an extend declaration, we require `A` is

Reviewer 2: This one was already using the new formulation, so has been switched around one too many or one too few times :)

```suggestion
This requirement functions recursively. Given an interface `B` that extends
another interface `A`: By naming `A` in an extend declaration, we require `A` is
```
